### PR TITLE
Issue #243: add unit tests for AuthenticationController, AuthenticationSDJpaService, UserSDJpaService (method findUserByEmail)

### LINKED ISSUE #243: Unit tests for new login endpoint logic
We need to write tests for https://github.com/jmprathab/MyHome/pull/240/files.

Classes that should be covered:
- AuthenticationController.java
- AuthenticationSDJpaService.java 
- UserSDJpaService.java (method findUserByEmail)

### PR DESCRIPTION
**Pull request by hsergiu**
## 🚀 Description
Created multiple tests for each method. Only a single one for login() in AuthenticationController.

## 📄 Motivation and Context

Closes: #243 

## 📦 Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] My code follows the code style of this project(Do your best to follow code styles. If none apply just skip this).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: service/src/test/java/com/myhome/services/unit/UserSDJpaServiceTest.java [hunk 1]
```diff
@@ -44,17 +44,10 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertNotEquals;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/services/unit/UserSDJpaServiceTest.java]: Wildcards imports.
Why is it bad? Read it here https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

How you can fix it in intellij?
https://stackoverflow.com/a/25916603/5785871

File: service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+  @Test
+  void loginUserNotFound(){
+    // given
+    LoginRequest request = getDefaultLoginRequest();
+    given(userSDJpaService.findUserByEmail(request.getEmail()))
+        .willReturn(Optional.empty());
+
+    // when and then
+    Assertions.assertThrows(UserNotFoundException.class, () -> authenticationSDJpaService.login(request));
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java]: Could you use static import for `assertThrows`?

File: service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    // given
+    LoginRequest request = getDefaultLoginRequest();
+    UserDto userDto = getDefaultUserDtoRequest();
+    given(userSDJpaService.findUserByEmail(request.getEmail()))
+        .willReturn(Optional.of(userDto));
+    given(passwordEncoder.matches(request.getPassword(), userDto.getEncryptedPassword()))
+        .willReturn(false);
+
+    // when and then
+    Assertions.assertThrows(CredentialsIncorrectException.class, () -> authenticationSDJpaService.login(request));
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java]: Same as above - static import please

File: service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+  }
+  private LoginRequest getDefaultLoginRequest() {
+    LoginRequest request = new LoginRequest();
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java]: Generated classes (in this case `LoginRequest`) have something similar to builder pattern.
You can write `LoginRequest.email(USER_EMAIL).password(REQUEST_PASSWORD)`

File: service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  @Test
+  void loginSuccess(){
+    // given
+    LoginRequest request = getDefaultLoginRequest();
+    UserDto userDto = getDefaultUserDtoRequest();
+    AppJwt appJwt = getDefaultJwtToken(userDto);
+    String encodedJwt = appJwtEncoderDecoder.encode(appJwt, SECRET);
+    ReflectionTestUtils.setField(authenticationSDJpaService, "tokenExpirationTime", Duration.ofDays(1));
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/services/unit/AuthenticationSDJpaServiceTest.java]: Using reflection is a bad practice.
We can make `AllArgsConstructor` annotation above `AuthenticationSDJpaService` or create all args constructor by hand and use there the annotations:
```
public AuthenticationSDJpaService(@Value("${token.expiration_time}") Duration tokenExpirationTime,
                                    @Value("${token.secret}") String tokenSecret,
                                    UserSDJpaService userSDJpaService,
                                    AppJwtEncoderDecoder appJwtEncoderDecoder,
                                    PasswordEncoder passwordEncoder) {
    this.tokenExpirationTime = tokenExpirationTime;
    this.tokenSecret = tokenSecret;
    this.userSDJpaService = userSDJpaService;
    this.appJwtEncoderDecoder = appJwtEncoderDecoder;
    this.passwordEncoder = passwordEncoder;
  }
```

That will give you a way to not inject properties with reflection :)

File: service/src/test/java/com/myhome/controllers/AuthenticationControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        LoginRequest loginRequest = getDefaultLoginRequest();
+        AuthenticationData authenticationData = getDefaultAuthenticationData();
+        given(authenticationService.login(loginRequest))
+                .willReturn(authenticationData);
+        // when
+        ResponseEntity<Void> response = authenticationController.login(loginRequest);
+
+        // then
+        assertEquals(HttpStatus.OK, response.getStatusCode());
+        assertEquals(response.getHeaders().size(), 2);
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/controllers/AuthenticationControllerTest.java]: Can we expand this assertion to check not just the size of headers but if there are proper headers with appropriate values (propagated from mocked authentication service)?

File: service/src/test/java/com/myhome/controllers/AuthenticationControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (mslowiak) [inline service/src/test/java/com/myhome/controllers/AuthenticationControllerTest.java]: That copyright can be deleted :D

**Review by mslowiak — CHANGES_REQUESTED** [review]
@hsergiu 
Thank you for your contribution.
There are few changes that I mentioned in the comments. Please reformat also all classes that were touched by you (I saw some code snippets that had no space after brackets etc).

File: service/src/main/java/com/myhome/services/springdatajpa/AuthenticationSDJpaService.java [hunk 8]
```diff
@@ -10,26 +10,36 @@
...
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
 
 @Slf4j
-@RequiredArgsConstructor
 @Service
 public class AuthenticationSDJpaService implements AuthenticationService {
 
   @Value("${token.expiration_time}")
```
Comment (mslowiak) [inline service/src/main/java/com/myhome/services/springdatajpa/AuthenticationSDJpaService.java]: If you added the `@Value` annotation in the constructor, then it doesn't need to be on the variable definition as well.

**Comment by mslowiak** [comment]
@hsergiu Just one more comment about this PR and we can proceed with merge :)

**Comment by hsergiu** [comment]
Thank you for the review. I made the changes.

**Review by mslowiak — APPROVED** [review]

**Pull request merged**